File: LeetCodePractice/src/reservoirSampling/ShuffleAnArray.java
package reservoirSampling;

import java.util.Random;

/*
Shuffle a set of numbers without duplicates.

Example:

// Init an array with set 1, 2, and 3.
int[] nums = {1,2,3};
Solution solution = new Solution(nums);

// Shuffle the array [1,2,3] and return its result. Any permutation of [1,2,3] must equally likely to be returned.
solution.shuffle();

// Resets the array back to its original configuration [1,2,3].
solution.reset();

// Returns the random shuffling of array [1,2,3].
solution.shuffle();

analysis:
Fisher–Yates shuffle
*/
public class ShuffleAnArray {
	private int[] nums;
	private Random random;
 	public ShuffleAnArray(int[] nums){
		this.nums = nums;
		random = new Random();
	}
	
	/** Resets the array to its original configuration and return it. */
    public int[] reset() {
        return nums;
    }
    
    /** Returns a random shuffling of the array. */
    public int[] shuffle() {
    	if(nums == null){
    		return null;
    	}
    	int[] res = nums.clone();
        int len = res.length;
        for(int i = 0; i < len; i++){
        	int index = randRange(i, len);
        	swap(res, i, index);
        }
        return res;
    }
    
    private int randRange(int min, int max){
    	return random.nextInt(max - min) + min;
    }

    private void swap(int[] arr, int i, int j){
    	int temp = arr[i];
    	arr[i] = arr[j];
    	arr[j] = temp;
    }
}
/**
 * Your Solution object will be instantiated and called as such:
 * Solution obj = new Solution(nums);
 * int[] param_1 = obj.reset();
 * int[] param_2 = obj.shuffle();
 */

File: LeetCodePython/ShuffleAnArray.py
"""
Given an integer array nums, design an algorithm to randomly shuffle the array. All permutations of the array should be equally likely as a result of the shuffling.

Implement the Solution class:

Solution(int[] nums) Initializes the object with the integer array nums.
int[] reset() Resets the array to its original configuration and returns it.
int[] shuffle() Returns a random shuffling of the array.

Example 1:
Input
["Solution", "shuffle", "reset", "shuffle"]
[[[1, 2, 3]], [], [], []]
Output
[null, [3, 1, 2], [1, 2, 3], [1, 3, 2]]

Explanation
Solution solution = new Solution([1, 2, 3]);
solution.shuffle();    // Shuffle the array [1,2,3] and return its result.
                       // Any permutation of [1,2,3] must be equally likely to be returned.
                       // Example: return [3, 1, 2]
solution.reset();      // Resets the array back to its original configuration [1,2,3]. Return [1, 2, 3]
solution.shuffle();    // Returns the random shuffling of array [1,2,3]. Example: return [1, 3, 2]

Constraints:
1 <= nums.length <= 50
-106 <= nums[i] <= 106
All the elements of nums are unique.
At most 104 calls in total will be made to reset and shuffle.

hint:
1 The solution expects that we always use the original array to shuffle() else some of the test cases fail.
"""
import random
from typing import List


class ShuffleAnArray:

    def __init__(self, nums: List[int]):
        self.nums = nums


    def reset(self) -> List[int]:
        return self.nums

    def shuffle(self) -> List[int]:
        n = len(self.nums)
        res = self.nums[:]
        for i in range(n):
            j = random.randrange(i, n)
            res[i], res[j] = res[j], res[i]
        return res



# Your Solution object will be instantiated and called as such:
# obj = Solution(nums)
# param_1 = obj.reset()
# param_2 = obj.shuffle()